src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IAsyncEnumerableOfTConverter.cs

```diff
@@ -0,0 +1,144 @@
+                {
+                    // converter was previously suspended due to a pending MoveNextAsync() task
+                    Debug.Assert(state.PendingTask is Task<bool> && state.PendingTask.IsCompleted);
+                    moveNextTask = new ValueTask<bool>((Task<bool>)state.PendingTask);
+                    state.Current.AsyncEnumeratorIsPendingCompletion = false;
+                    state.PendingTask = null;
+                }
+                else
+                {
+                    // converter was suspended for a different reason;
+                    // the last MoveNextAsync() call can only have completed with 'true'.
+                    moveNextTask = new ValueTask<bool>(true);
+                }
+            }
+
+            JsonConverter<TElement> converter = GetElementConverter(ref state);
+
+            // iterate through the enumerator while elements are being returned synchronously
+            for (; moveNextTask.IsCompleted; moveNextTask = enumerator.MoveNextAsync())
+            {
+                if (!moveNextTask.Result)
+                {
+                    return true;
+                }
+
+                if (ShouldFlush(writer, ref state))
+                {
+                    return false;
+                }
+
+                TElement element = enumerator.Current;
+                if (!converter.TryWrite(writer, element, options, ref state))
+                {
+                    return false;
+                }
+            }
+
+            // we have a pending MoveNextAsync() call;
+            // wrap inside a regular task so that it can be awaited multiple times;
+            // mark the current stackframe as pending completion.
+            Debug.Assert(state.PendingTask is null);
+            state.PendingTask = moveNextTask.AsTask();
+            state.Current.AsyncEnumeratorIsPendingCompletion = true;
+            return false;
+        }
+
+        internal override bool OnTryRead(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options, ref ReadStack state, out TAsyncEnumerable value)
+        {
+            if (!typeToConvert.IsAssignableFrom(typeof(IAsyncEnumerable<TElement>)))
+            {
+                ThrowHelper.ThrowNotSupportedException_CannotPopulateCollection(TypeToConvert, ref reader, ref state);
+            }
+
+            return base.OnTryRead(ref reader, typeToConvert, options, ref state, out value!);
+        }
+
+        protected override void Add(in TElement value, ref ReadStack state)
+        {
+            ((BufferedAsyncEnumerable)state.Current.ReturnValue!)._buffer.Add(value);
+        }
+
+        protected override void CreateCollection(ref Utf8JsonReader reader, ref ReadStack state, JsonSerializerOptions options)
+        {
+            state.Current.ReturnValue = new BufferedAsyncEnumerable();
+        }
+
+        private class BufferedAsyncEnumerable : IAsyncEnumerable<TElement>
```

Comment:
This could also be done as an iterator (you might need to suppress an erroneous warning about lack of awaits):
```C#
private static async IAsyncEnumerable<TElement> YieldAsync(List<TElement> buffer)
{
    for (int i = 0; i < buffer.Length; i++)
        yield return buffer[i];
}
```
Less code, and it would save on an allocation for the enumerator in the common case.

Reply:
Can we really avoid the enumerator allocation though? It would need to be written as follows:
```csharp
internal class BufferedEnumerator : IAsyncEnumerable<T>
{
    public readonly List<T> _buffer = new();
    public async IAsyncEnumerator<T> GetEnumerator(CancellationToken _)
    {
        for (int i = 0; i < _buffer.Length; i++)
            yield return _buffer[i];
    }
}
```

Reply:
> It would need to be written as follows:

Why does it need to be an iterator returning an enumerator rather than enumerable?

Reply:
I see, because you want to be able to store the one object into a "ReturnValue" field that will both be returned and used to access the List it stores?

In that case, I'd be inclined to still use the compiler generated implementation for the IAsyncEnumerator.  Less code to maintain, less code to get wrong, and it seems unlikely to be an issue perf-wise, given the use case.